### ISSUE TITLE
Stack overflow for global tool on OSX

### ISSUE THREAD

**Issue by theimowski** [fsprojects/fantomas#591]
### Description

I have a 1000 LOC file that I just added a couple more lines to and then Fantomas started throwing "Stack overflow" on my MacOS. Same file gets formatted ok in fantomas-ui. It works fine also if I remove any single instruction / expression from the code.

Maybe it's just a matter of configuring some environmental thing for dotnet core to extend stack size? 

EDIT:
Tried `ulimit -s 65532` as per https://stackoverflow.com/a/13261334/1397724 but without effect

**Comment by nojaf** [comment]
Hey @theimowski , could you show us that file maybe?
It is very hard to say something useful without knowing what we are dealing with here 😅.

Fantomas does sometimes format blocks multiple times and we might have a memory leak there.

**Comment by theimowski** [comment]
I'm afraid I can't, tried to come up with a minimal repro sample, but as mentioned above - whenever I remove a single line of code here or there it works, so it's hard for me to assess which part is causing the issue.

Unfortunately I know this doesn't help much but hoped that tweaking some configuration could help

**Comment by theimowski** [comment]
It's happening when I use custom `pageWidth` below 100, setting it to 100 or above makes the issue go away.

I'll try to debug this next week

**Comment by theimowski** [comment]
Working on chasing this down. 
The file is a fable elmish app - I have a `Msg` type with 30 union cases, and then in `update` function I have a pattern matching on that `Msg`. 

What I found (not sure if a good trace) is that when I removed 5 cases (leaving pattern matching with 25 cases) from the pattern matching expression, I got rid of the stack overflow.

So I run a debugging session with the 25-case pattern matching to see what the call stack looks like when handling last clause, maybe this helps:

![Screenshot 2019-12-11 at 09 56 11](https://user-images.githubusercontent.com/5909015/70606555-030bc400-1bfd-11ea-829d-da862a223d86.png)
![Screenshot 2019-12-11 at 09 56 26](https://user-images.githubusercontent.com/5909015/70606562-069f4b00-1bfd-11ea-8db8-570463f07241.png)
![Screenshot 2019-12-11 at 09 56 40](https://user-images.githubusercontent.com/5909015/70606565-0901a500-1bfd-11ea-87f2-3135ef96fbd0.png)
